### PR TITLE
Factor expressions scripts out to lang-expression plugin

### PR DESCRIPTION
**Pull request by rmuir**
This is part of moving script _impls_ out to plugins.

See https://github.com/elastic/elasticsearch/issues/13725 for the motivation.

This one is a pretty easy win, it means removing conflicts for antlr, asm, asm-commons, etc for users, and is a step towards containing classloader security.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 1]
```diff
@@ -464,152 +464,148 @@ public String toString() {
...
-        exceptions.put(org.elasticsearch.index.engine.EngineClosedException.class, 131);
-        exceptions.put(org.elasticsearch.action.NoShardAvailableActionException.class, 132);
-        exceptions.put(org.elasticsearch.action.UnavailableShardsException.class, 133);
-        exceptions.put(org.elasticsearch.index.engine.FlushFailedEngineException.class, 134);
-        exceptions.put(org.elasticsearch.common.breaker.CircuitBreakingException.class, 135);
-        exceptions.put(org.elasticsearch.transport.NodeNotConnectedException.class, 136);
-        exceptions.put(org.elasticsearch.index.mapper.StrictDynamicMappingException.class, 137);
-        exceptions.put(org.elasticsearch.action.support.replication.TransportReplicationAction.RetryOnReplicaException.class, 138);
-        exceptions.put(org.elasticsearch.indices.TypeMissingException.class, 139);
+        exceptions.put(org.elasticsearch.index.snapshots.IndexShardSnapshotFailedException.class, exceptions.size());
```
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Although I understand the goal here, I think this should go in another PR.
We don't clearly see here what has been removed if anything.

Also, it might mean that a exception ID can change between two versions (3.0.0 to 3.0.1)  and I guess that it could fail the deserialization when an exception is sent between nodes with different versions

File: plugins/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionPlugin.java [hunk 2]
```diff
@@ -19,24 +19,22 @@
...
 
-/**
- * Exception used to wrap exceptions occuring while running expressions.
- */
-public class ExpressionScriptExecutionException extends ElasticsearchException {
-    public ExpressionScriptExecutionException(String msg, Throwable cause) {
-        super(msg, cause);
+    @Override
+    public String name() {
+        return "lang-expression";
```
Comment (dadoonet) [inline plugins/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionPlugin.java]: Not related at all but I think we should change names to "script-expression" (same for other "lang-" plugins).
We should do that in another PR for 3.0

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 1 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: OK. i can add the numbers back, it does not matter to me. personally I was thinking, lets clean house on these exceptions anyway.

What i did was remove the two expression-specific exceptions. There is just no need for that. I use the elasticsearch ScriptException everywhere instead. and maybe later it should just be javax.script.ScriptException.

File: plugins/lang-expression/src/test/java/org/elasticsearch/script/expression/MoreExpressionTests.java [hunk 3]
```diff
@@ -60,7 +63,13 @@
 import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.notNullValue;
 
-public class ExpressionScriptIT extends ESIntegTestCase {
+// TODO: please convert to unit tests!
```
Comment (dadoonet) [inline plugins/lang-expression/src/test/java/org/elasticsearch/script/expression/MoreExpressionTests.java]: ++

**Comment by dadoonet** [comment]
I left unrelated comments mostly. I think it's good but I think we should have another reviewer.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 1 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: You still wont be able to see the changes, because i must renumber every element. that is how the old system here worked.

**Comment by rjernst** [comment]
I fixed some of the remaining tests to use mocks, so further work pulling out script engines should be a little bit easier. LGTM.

**Comment by s1monw** [comment]
LGTM 2

**Comment by rmuir** [comment]
I just backported this to 2.x
